File: 2017/fall/3/binary1.c
// Recursive binary search

#include <cs50.h>
#include <stdio.h>
#include <string.h>

// Names in a phone book
string book[] = {
    "Chen",
    "Kernighan",
    "Leitner",
    "Lewis",
    "Malan",
    "Muller",
    "Seltzer",
    "Shieber",
    "Smith"
};

int main(void)
{
    // Prompt user for name
    string name = get_string("Name: ");

    // Search for name
    int left = 0, right = sizeof(book) / sizeof(string) - 1;
    while (left <= right)
    {
        // Look at middle
        int middle = (left + right) / 2;
        if (strcmp(name, book[middle]) == 0)
        {
            printf("Calling %s\n", name);
            return 0;
        }

        // Search left half
        else if (strcmp(name, book[middle]) < 0)
        {
            right = middle - 1;
        }

        // Search right half
        else if (strcmp(name, book[middle]) > 0)
        {
            left = middle + 1;
        }
    }
    printf("Quitting\n");
    return 1;
}

File: 2017/fall/8/src8/weeks/3/binary1.py
# Recursive binary search

from cs50 import get_string

# Names in a phone book
book = [
    "Chen",
    "Kernighan",
    "Leitner",
    "Lewis",
    "Malan",
    "Muller",
    "Seltzer",
    "Shieber",
    "Smith"]


def main():

    # Prompt user for name
    name = get_string("Name: ")

    # Search for name
    if search(name, book):
        print(f"Calling {name}")
    else:
        print("Quitting")


def search(name, names):
    """Search names for name"""

    # No more names to search
    if not names:
        return False

    # Look at middle
    middle = len(names) // 2
    if name == names[middle]:
        return True

    # Search left half
    elif name < names[middle]:
        return search(name, names[:middle])

    # Search right half
    elif name > names[middle]:
        return search(name, names[middle + 1:])


if __name__ == "__main__":
    main()
